typescript/web/src/components/app-lifecycle-manager/welcome-modal/welcome-modal.tsx

```diff
@@ -1,185 +1,357 @@
+
+    setIsLoadingWorkerAndDemo(true);
+
+    try {
+      await checkServiceWorkerReady();
+    } catch (e) {
+      if (e.message === messageNoWindow) {
+        // In Next JS SSR, the window is not available.
+        return;
+      }
+      if (process.env.STORYBOOK) {
+        // In Storybook, the Service Worker is not available.
+        return;
+      }
+      throw e;
+    }
+    setIsServiceWorkerActive(true);
+
+    const { data: getDatasetsResult } = await client.query({
+      query: getDatasetsQuery,
+    });
+
+    const demoDataset =
+      getDatasetsResult?.datasets == null
+        ? undefined
+        : getDatasetsResult?.datasets.filter(
+            (dataset: DatasetType) => dataset.name === "Demo dataset"
+          )?.[0] ?? undefined;
+
+    if (!demoDataset) {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const { data: createDemoDatasetResult, errors: createDemoDatasetErrors } =
+        await client.mutate({
+          mutation: createDemoDatasetQuery,
+          refetchQueries: [
+            "getDatasets",
+            "getDatasetById",
+            "getAllImagesOfADataset",
+            "getDataset",
+            "countLabelsOfDataset",
+            "getLabelClassesOfDataset",
+            "image",
+            "getImageLabels",
+          ],
+        });
+      if (createDemoDatasetErrors) {
+        throw createDemoDatasetErrors[0];
+      }
+    }
+
+    await router.prefetch(tutorialDatasetFirstImageUrl);
+
+    setIsLoadingWorkerAndDemo(false);
+  } catch (error) {
+    setBrowserError(error);
+  }
+};
+
+export const WelcomeModal = ({
+  initialIsServiceWorkerActive = false,
+  initialBrowserWarning = false,
+  initialIsLoadingWorkerAndDemo = false,
+  initialBrowserError = undefined,
+}: {
+  initialIsServiceWorkerActive?: boolean;
+  initialBrowserWarning?: boolean;
+  initialIsLoadingWorkerAndDemo?: boolean;
+  initialBrowserError?: Error;
+}) => {
+  const router = useRouter();
+
+  const handleError = useErrorHandler();
+  const client = useApolloClient();
+
+  const [{ hasUserTriedApp }, setHasUserTriedApp] = useCookies([
+    "hasUserTriedApp",
+  ]);
+  const [{ tryDespiteBrowserWarning }, setTryDespiteBrowserWarning] =
+    useCookies(["tryDespiteBrowserWarning"]);
+
+  const [browserWarning] = useState(() => {
+    const name = browser?.name;
+    const os = browser?.os;
+    const versionNumber = parseInt(browser?.version ?? "0", 10);
+    if (
+      initialBrowserWarning ||
+      !(
+        (name === "firefox" && versionNumber >= 44) ||
+        (name === "edge-chromium" && versionNumber >= 44) ||
+        (name === "chrome" && versionNumber >= 45) ||
+        (name === "safari" && versionNumber >= 14)
+      ) ||
+      os === "BlackBerry OS" ||
+      os === "Windows Mobile" ||
+      os === "Windows 3.11" ||
+      os === "Windows 95" ||
+      os === "Windows 98" ||
+      os === "Windows 2000" ||
+      os === "Windows ME"
```

Comment:
Small proposal, feel free to ignore

You could have done something like: 
```typescript
const unsupportedOSs = ["BlackBerry OS", /* ... */ , "Windows ME"];
const isOSSupported = unsupportedOSs.includes(os);
```

Reply:
Yes. I could have written it that way, your solution is good. 

But I'll ignore it for now since that would add some indirection. Here the condition is obvious, don't have to scroll on top of the file to have the list of unsupported stuff.